File: api/app/services/capsules.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Any
from uuid import uuid4

from fastapi import HTTPException, status

from ..models import CreateCapsuleRequest
from ..supabase_gateway import SupabaseError, SupabaseGateway

ALLOWED_PHOTO_TYPES = {
    "image/jpeg": "jpg",
    "image/png": "png",
    "image/webp": "webp",
    "image/heic": "heic",
    "image/heif": "heif",
}


@dataclass(frozen=True)
class PhotoUpload:
    content: bytes
    content_type: str
    width: int | None
    height: int | None


def capsule_item(
    capsule_id: str,
    item_type: str,
    position: int,
    *,
    body: str | None = None,
    storage_path: str | None = None,
    metadata: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Build a uniform row for PostgREST bulk inserts."""
    return {
        "capsule_id": capsule_id,
        "item_type": item_type,
        "position": position,
        "body": body,
        "storage_path": storage_path,
        "metadata": metadata or {},
    }


def capsule_summary(row: dict[str, Any], user_id: str) -> dict[str, Any]:
    return {
        "id": str(row["id"]),
        "title": row["title"],
        "subtitle": row.get("subtitle") or "",
        "capsule_type": row["capsule_type"],
        "recipient": row["recipient"],
        "open_at": row["open_at"],
        "status": row["status"],
        "accent": row["accent"],
        "emoji": row["emoji"],
        "reminder_enabled": row["reminder_enabled"],
        "item_counts": row.get("item_counts") or {},
        "integrity_hash": row.get("seal_hash"),
        "opened_at": row.get("opened_at"),
        "created_at": row["created_at"],
        "collaborative": bool(row.get("collaborative", False)),
        "is_shared": str(row["user_id"]) != user_id,
        "owner_id": str(row["user_id"]),
        "is_owner": str(row["user_id"]) == user_id,
    }


def public_error(error: SupabaseError) -> HTTPException:
    message = error.message
    lowered = message.lower()
    if error.status_code in {401, 403}:
        return HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Access denied.")
    if error.status_code == 404:
        return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=message)
    if "not ready" in lowered or "unlock" in lowered:
        return HTTPException(status_code=status.HTTP_423_LOCKED, detail=message)
    if error.status_code in {400, 409} or error.code in {"23505", "23514"}:
        return HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=message)
    return HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=message)


class CapsuleService:
    def __init__(self, gateway: SupabaseGateway) -> None:
        self.gateway = gateway

    async def list(self, token: str, user_id: str) -> list[dict[str, Any]]:
        try:
            rows = await self.gateway.list_capsules(token)
        except SupabaseError as error:
            raise public_error(error) from error
        return [capsule_summary(row, user_id) for row in rows]

    async def create(
        self,
        token: str,
        user_id: str,
        payload: CreateCapsuleRequest,
        photos: list[PhotoUpload],
    ) -> dict[str, Any]:
        if payload.open_at <= datetime.now(UTC) + timedelta(minutes=10):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Choose an unlock time at least 10 minutes in the future.",
            )
        if len(payload.photos) != len(photos):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Photo metadata does not match the uploaded files.",
            )

        capsule_id: str | None = None
        uploaded_paths: list[str] = []
        try:
            draft = await self.gateway.create_draft(
                token,
                user_id,
                {
                    "title": payload.title,
                    "subtitle": payload.subtitle,
                    "capsule_type": payload.capsule_type,
                    "recipient": payload.recipient,
                    "open_at": payload.open_at.isoformat(),
                    "accent": payload.accent,
                    "emoji": payload.emoji,
                    "reminder_enabled": payload.reminder_enabled,
                    "collaborative": payload.collaborative,
                },
            )
            capsule_id = str(draft["id"])

            for username in payload.shared_with_usernames:
                await self.gateway.invite_member(token, capsule_id, username)

            items: list[dict[str, Any]] = []
            position = 0
            if payload.letter:
                items.append(
                    capsule_item(
                        capsule_id,
                        "letter",
                        position,
                        body=payload.letter,
                    )
                )
                position += 1
            for goal in payload.goals:
                items.append(
                    capsule_item(
                        capsule_id,
                        "goal",
                        position,
                        body=goal.text,
                        metadata={"completed": goal.completed},
                    )
                )
                position += 1
            for prediction in payload.predictions:
                items.append(
                    capsule_item(
                        capsule_id,
                        "prediction",
                        position,
                        body=prediction,
                    )
                )
                position += 1
            for metadata, upload in zip(payload.photos, photos, strict=True):
                extension = ALLOWED_PHOTO_TYPES[upload.content_type]
                path = f"{user_id}/{capsule_id}/{uuid4()}.{extension}"
                await self.gateway.upload_media(token, path, upload.content, upload.content_type)
                uploaded_paths.append(path)
                items.append(
                    capsule_item(
                        capsule_id,
                        "photo",
                        position,
                        storage_path=path,
                        metadata={
                            "client_id": metadata.id,
                            "width": upload.width,
                            "height": upload.height,
                        },
                    )
                )
                position += 1

            await self.gateway.insert_items(token, items)
            if payload.collaborative:
                draft["item_counts"] = {
                    "letter": 1 if payload.letter else 0,
                    "goals": len(payload.goals),
                    "predictions": len(payload.predictions),
                    "photos": len(payload.photos),
                }
                return capsule_summary(draft, user_id)
            sealed = await self.gateway.seal_capsule(token, capsule_id)
            return capsule_summary(sealed, user_id)
        except SupabaseError as error:
            if capsule_id:
                try:
                    await self.gateway.remove_media(token, uploaded_paths)
                except SupabaseError:
                    pass
                try:
                    await self.gateway.delete_draft(token, capsule_id)
                except SupabaseError:
                    pass
            raise public_error(error) from error

    async def draft(self, token: str, user_id: str, capsule_id: str) -> dict[str, Any]:
        try:
            capsule = await self.gateway.get_capsule(token, capsule_id)
            if capsule["status"] != "draft" or not capsule.get("collaborative", False):
                raise HTTPException(
                    status_code=status.HTTP_409_CONFLICT,
                    detail="This capsule is no longer accepting contributions.",
                )
            rows = await self.gateway.list_draft_items(token, capsule_id)
            contributor_ids = list({str(row["contributor_id"]) for row in rows})
            profiles = await self.gateway.get_profiles_by_id(token, contributor_ids)
        except SupabaseError as error:
            raise public_error(error) from error
        by_id = {str(profile["id"]): profile for profile in profiles}
        items = []
        for row in rows:
            item = {
                "id": str(row["id"]),
                "item_type": row["item_type"],
                "position": row["position"],
                "body": row.get("body"),
                "metadata": row.get("metadata") or {},
                "contributor": by_id.get(str(row["contributor_id"])),
            }
            if row["item_type"] == "photo":
                item["media_path"] = f"/v1/capsules/{capsule_id}/media/{row['id']}"
            items.append(item)
        return {"capsule": capsule_summary(capsule, user_id), "items": items}

    async def contribute(
        self, token: str, user_id: str, capsule_id: str, body: str
    ) -> dict[str, Any]:
        try:
            row = await self.gateway.add_contribution(token, capsule_id, body)
            profile = await self.gateway.get_profile(token, user_id)
        except SupabaseError as error:
            raise public_error(error) from error
        return {
            "id": str(row["id"]),
            "item_type": row["item_type"],
            "position": row["position"],
            "body": row.get("body"),
            "metadata": row.get("metadata") or {},
            "contributor": profile,
        }

    async def seal(self, token: str, user_id: str, capsule_id: str) -> dict[str, Any]:
        try:
            row = await self.gateway.seal_capsule(token, capsule_id)
        except SupabaseError as error:
            raise public_error(error) from error
        return capsule_summary(row, user_id)

    async def reveal(self, token: str, user_id: str, capsule_id: str) -> dict[str, Any]:
        try:
            payload = await self.gateway.reveal_capsule(token, capsule_id)
            contributor_ids = list(
                {
                    str(item["contributor_id"])
                    for item in (payload.get("items") or [])
                    if item.get("contributor_id")
                }
            )
            profiles = await self.gateway.get_profiles_by_id(token, contributor_ids)
        except SupabaseError as error:
            raise public_error(error) from error

        capsule = capsule_summary(payload["capsule"], user_id)
        profiles_by_id = {str(profile["id"]): profile for profile in profiles}
        items = []
        for item in payload.get("items") or []:
            public_item = {
                "id": str(item["id"]),
                "item_type": item["item_type"],
                "position": item["position"],
                "body": item.get("body"),
                "metadata": item.get("metadata") or {},
                "contributor": profiles_by_id.get(str(item.get("contributor_id"))),
            }
            if item["item_type"] == "photo":
                public_item["media_path"] = f"/v1/capsules/{capsule_id}/media/{item['id']}"
            items.append(public_item)
        return {
            "capsule": capsule,
            "items": items,
            "trusted_time": payload["trusted_time"],
        }

File: mobile/src/api/capsules.ts
import { Platform } from "react-native";
import {
  Capsule,
  CapsuleItemCounts,
  CapsulePhoto,
  CapsuleContribution,
  CapsuleType,
  GoalItem,
  NewCapsule,
} from "../types";
import { apiUrl } from "./client";
import { authorizedRequest, currentSession } from "./session";

type ApiCapsule = {
  id: string;
  title: string;
  subtitle: string;
  capsule_type: CapsuleType;
  recipient: string;
  open_at: string;
  status: "draft" | "sealed" | "opened";
  accent: string;
  emoji: string;
  reminder_enabled: boolean;
  item_counts: Partial<CapsuleItemCounts>;
  integrity_hash: string | null;
  opened_at: string | null;
  created_at: string;
  is_shared: boolean;
  owner_id: string;
  is_owner: boolean;
  collaborative: boolean;
};

type ApiItem = {
  id: string;
  item_type: "letter" | "goal" | "prediction" | "photo";
  position: number;
  body: string | null;
  metadata: Record<string, unknown>;
  media_path?: string;
  contributor?: { id: string; display_name: string; username: string };
};

type RevealResponse = {
  data: { capsule: ApiCapsule; items: ApiItem[]; trusted_time: string };
  authority: "database";
};

const EMPTY_COUNTS: CapsuleItemCounts = {
  letter: 0,
  goals: 0,
  predictions: 0,
  photos: 0,
};

function normalizeCounts(value: Partial<CapsuleItemCounts> | null): CapsuleItemCounts {
  return {
    letter: Number(value?.letter ?? 0),
    goals: Number(value?.goals ?? 0),
    predictions: Number(value?.predictions ?? 0),
    photos: Number(value?.photos ?? 0),
  };
}

function mapCapsule(row: ApiCapsule): Capsule {
  return {
    id: row.id,
    remoteId: row.id,
    title: row.title,
    subtitle: row.subtitle,
    type: row.capsule_type,
    recipient: row.recipient,
    letter: "",
    goals: [],
    predictions: [],
    photos: [],
    createdAt: row.created_at,
    unlockAt: row.open_at,
    openedAt: row.opened_at ?? undefined,
    status: row.status,
    accent: row.accent,
    emoji: row.emoji,
    reminderEnabled: row.reminder_enabled,
    integrityHash: row.integrity_hash ?? undefined,
    syncStatus: "synced",
    contentLoaded: false,
    itemCounts: normalizeCounts(row.item_counts),
    isShared: row.is_shared,
    ownerId: row.owner_id,
    isOwner: row.is_owner,
    collaborative: row.collaborative,
    contributions: [],
  };
}

function extensionForUri(uri: string) {
  const match = /\.([a-zA-Z0-9]+)(?:\?|$)/.exec(uri);
  const extension = match?.[1]?.toLowerCase();
  return extension && ["jpg", "jpeg", "png", "webp", "heic", "heif"].includes(extension)
    ? extension
    : "jpg";
}

function mimeForExtension(extension: string) {
  if (extension === "png") return "image/png";
  if (extension === "webp") return "image/webp";
  if (extension === "heic") return "image/heic";
  if (extension === "heif") return "image/heif";
  return "image/jpeg";
}

export async function listCapsules() {
  const response = await authorizedRequest<{ data: ApiCapsule[] }>("/v1/capsules");
  return response.data.map(mapCapsule);
}

export async function createCapsule(input: NewCapsule) {
  const form = new FormData();
  form.append(
    "payload",
    JSON.stringify({
      title: input.title,
      subtitle: input.subtitle,
      capsule_type: input.type,
      recipient: input.recipient,
      letter: input.letter,
      goals: input.goals,
      predictions: input.predictions,
      photos: input.photos.map(({ id, width, height }) => ({ id, width, height })),
      open_at: input.unlockAt,
      accent: input.accent,
      emoji: input.emoji,
      reminder_enabled: input.reminderEnabled,
      shared_with_usernames: input.sharedWithUsernames ?? [],
      collaborative: input.collaborative ?? false,
    }),
  );

  for (const photo of input.photos) {
    const extension = extensionForUri(photo.uri);
    const type = mimeForExtension(extension);
    const name = `${photo.id}.${extension}`;
    if (Platform.OS === "web") {
      const response = await fetch(photo.uri);
      form.append("photos", await response.blob(), name);
    } else {
      form.append("photos", { uri: photo.uri, type, name } as unknown as Blob);
    }
  }

  const response = await authorizedRequest<{ data: ApiCapsule }>("/v1/capsules", {
    method: "POST",
    body: form,
  });
  return mapCapsule(response.data);
}

export async function revealCapsule(capsuleId: string) {
  const response = await authorizedRequest<RevealResponse>(
    `/v1/capsules/${capsuleId}/reveal`,
    { method: "POST" },
  );
  const { capsule, items, trusted_time: trustedTime } = response.data;
  const goals: GoalItem[] = items
    .filter((item) => item.item_type === "goal" && item.body)
    .map((item) => ({
      id: item.id,
      text: item.body!,
      completed: item.metadata.completed === true,
    }));
  const predictions = items
    .filter((item) => item.item_type === "prediction" && item.body)
    .map((item) => item.body!);
  const photos: CapsulePhoto[] = items
    .filter((item) => item.item_type === "photo" && item.media_path)
    .map((item) => ({
      id: item.id,
      uri: apiUrl(item.media_path!),
      width: typeof item.metadata.width === "number" ? item.metadata.width : undefined,
      height: typeof item.metadata.height === "number" ? item.metadata.height : undefined,
      requiresAuth: true,
    }));
  const contributions: CapsuleContribution[] = items
    .filter((item) => item.metadata.collaborative === true && item.body)
    .map((item) => ({
      id: item.id,
      body: item.body!,
      contributor: item.contributor
        ? {
            id: item.contributor.id,
            displayName: item.contributor.display_name,
            username: item.contributor.username,
          }
        : undefined,
    }));
  return {
    ...mapCapsule(capsule),
    letter:
      items.find(
        (item) => item.item_type === "letter" && item.metadata.collaborative !== true,
      )?.body ?? "",
    goals,
    predictions,
    photos,
    contributions,
    status: "opened" as const,
    openedAt: capsule.opened_at ?? trustedTime,
    contentLoaded: true,
  };
}

export async function loadCollaborativeDraft(capsuleId: string) {
  const response = await authorizedRequest<{
    data: { capsule: ApiCapsule; items: ApiItem[] };
  }>(`/v1/capsules/${capsuleId}/draft`);
  const { capsule, items } = response.data;
  const collaborativeItems = items.filter((item) => item.metadata.collaborative === true);
  const contributions: CapsuleContribution[] = collaborativeItems
    .filter((item) => item.body)
    .map((item) => ({
      id: item.id,
      body: item.body!,
      contributor: item.contributor
        ? {
            id: item.contributor.id,
            displayName: item.contributor.display_name,
            username: item.contributor.username,
          }
        : undefined,
    }));
  const photos: CapsulePhoto[] = items
    .filter((item) => item.item_type === "photo" && item.media_path)
    .map((item) => ({
      id: item.id,
      uri: apiUrl(item.media_path!),
      width: typeof item.metadata.width === "number" ? item.metadata.width : undefined,
      height: typeof item.metadata.height === "number" ? item.metadata.height : undefined,
      requiresAuth: true,
    }));
  return {
    ...mapCapsule(capsule),
    letter:
      items.find(
        (item) => item.item_type === "letter" && item.metadata.collaborative !== true,
      )?.body ?? "",
    goals: items
      .filter((item) => item.item_type === "goal" && item.body)
      .map((item) => ({ id: item.id, text: item.body!, completed: false })),
    predictions: items
      .filter((item) => item.item_type === "prediction" && item.body)
      .map((item) => item.body!),
    photos,
    contributions,
    contentLoaded: true,
  };
}

export async function addCapsuleContribution(capsuleId: string, body: string) {
  const response = await authorizedRequest<{ data: ApiItem }>(
    `/v1/capsules/${capsuleId}/contributions`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ body }),
    },
  );
  const item = response.data;
  return {
    id: item.id,
    body: item.body ?? body,
    contributor: item.contributor
      ? {
          id: item.contributor.id,
          displayName: item.contributor.display_name,
          username: item.contributor.username,
        }
      : undefined,
  } satisfies CapsuleContribution;
}

export async function sealCollaborativeCapsule(capsuleId: string) {
  const response = await authorizedRequest<{ data: ApiCapsule }>(
    `/v1/capsules/${capsuleId}/seal`,
    { method: "POST" },
  );
  return mapCapsule(response.data);
}

export function mediaAuthorizationHeader() {
  const active = currentSession();
  return active ? { Authorization: `Bearer ${active.accessToken}` } : undefined;
}

export function countsForCapsule(capsule: Capsule): CapsuleItemCounts {
  return capsule.itemCounts ?? {
    ...EMPTY_COUNTS,
    letter: capsule.letter ? 1 : 0,
    goals: capsule.goals.length,
    predictions: capsule.predictions.length,
    photos: capsule.photos.length,
  };
}
